crypto/ec/ecx_meth.c

```diff
@@ -16,30 +16,88 @@
 #include "internal/evp_int.h"
 #include "ec_lcl.h"
 
-#define X25519_KEYLEN        32
 #define X25519_BITS          253
 #define X25519_SECURITY_BITS 128
 
 #define ED25519_SIGSIZE      64
 
-typedef struct {
-    unsigned char pubkey[X25519_KEYLEN];
-    unsigned char *privkey;
-} X25519_KEY;
+#define X448_BITS            448
+#define ED448_BITS           456
+#define X448_SECURITY_BITS   224
+
+#define ED448_SIGSIZE        114
+
+#define ISX448(id)      ((id) == EVP_PKEY_X448 ? 1 : 0)
+#define IS25519(id)     (((id) == EVP_PKEY_X25519 || (id) == EVP_PKEY_ED25519) \
+                            ? 1 : 0)
+#define KEYLENID(id)    (IS25519(id) ? X25519_KEYLEN \
+                                     : ((id) == EVP_PKEY_X448 ? X448_KEYLEN \
+                                                              : ED448_KEYLEN))
+#define KEYLEN(p)       KEYLENID((p)->ameth->pkey_id)
+
+#define MAX_KEYLEN  ED448_KEYLEN
+
+static ossl_inline void getprivkey(unsigned char **priv, const EVP_PKEY *pk)
+{
+    if (IS25519(pk->ameth->pkey_id)) {
+        if (pk->pkey.curve25519 == NULL)
+            return;
+        *priv = pk->pkey.curve25519->privkey;
+    } else if(ISX448(pk->ameth->pkey_id)) {
+        if (pk->pkey.x448 == NULL)
+            return;
+        *priv = pk->pkey.x448->privkey;
+    } else {
```

Comment:
I'm not sure how comfortable I am with an unqualified 'else' here (and similarly, later).

Reply:
It really shouldn't be anything other than ED448 here. I could put an assert in if it makes you feel more comfortable? But I'm not sure whether it is worth it.

Reply:
I mean, yes, this is all internal and I trust you do have gotten things right.
And if we tried switching to a switch statement, we'd get our choice of compilers warning us for no 'default' label or for an unreachable default label (if we use an enum and use all values of it)...